### ISSUE TITLE
[BUG] Traces not sent with CLI scripts

### ISSUE THREAD

**Issue by bgaillard** [DataDog/dd-trace-php#981]
**Describe the bug**

Hi, i'm trying to instrument a very simple PHP CLI script (before integrating more complex CLI scripts later).

The script is the following.

```php
<?php
// test.php

declare(strict_types=1);
ini_set('display_errors', 'stderr');

function hello_world(array $message) {
    echo "Hello world !" . PHP_EOL;
}

hello_world(['key_0' => 'value_0']);
```

When I execute the PHP script I get the following output. You can observe that the traces are not sent after `Hello world !`. I tried this with `ddtrace` versions `0.41.1` and `0.47.1`.

Do you know why traces are not sent ? I though in CLI mode `ddtrace` instrumentation should work without any problem when correct environment variables are set. Did I miss something ?

```php
/tmp # DD_AGENT_HOST=datadog-agent DD_AGENT_PORT=8126 DD_TRACE_DEBUG=true DD_TRACE_ENABLED=true DD_TRACE_NO_AUTOLOADER=true DD_TRACE_CLI_ENABLED=true DD_TRACE_GENERATE_ROOT_SPAN=true DD_TRACE_AUTO_FLUSH_ENABLED=true DD_SERVICE=hello-world php test.php
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - New span test.php  recorded.
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Attempting integrations load
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration cakephp
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration curl
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration elasticsearch
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration eloquent
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration guzzle
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration laravel
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration lumen
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Integration memcached not available. New attempts WILL NOT be performed.
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Integration mongo not available. New attempts WILL NOT be performed.
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Integration mysqli not available. New attempts WILL NOT be performed.
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration pdo
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration predis
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration slim
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration symfony
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration web
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration zendframework
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration codeigniter
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration wordpress
[2020-08-04T17:00:57+00:00] [ddtrace] [debug] - Loaded integration yii
Hello world !
```

If I update my script to add a call to `\dd_trace_function` as described here https://github.com/DataDog/dd-trace-php/pull/834 then traces are correctly sent and I see the traces in Datadog.

```php
// test.php
...
\dd_trace_function('hello_world', function (\DDTrace\SpanData $span, array $args) {
    $span->name = 'hello_world';
    $span->service = 'hello-world';
    $span->resource = $args[0];

    echo "IN dd_trace_function." . PHP_EOL;
});

hello_world(['key_0' => 'value_0']);
```

Execution now displays logs showing that traces are sent.

```
/tmp #  DD_AGENT_HOST=datadog-agent DD_AGENT_PORT=8126 DD_TRACE_DEBUG=true DD_TRACE_ENABLED=true DD_TRACE_NO_AUTOLOADER=true DD_TRACE_CLI_ENABLED=true DD_TRACE_GENERATE_ROOT_SPAN=false DD_TRACE_AUTO_FLUSH_ENABLED=true DD_SERVICE_NAME=hello-world php test.php
...
Hello world !
IN dd_trace_function.
[2020-08-05T12:49:46+00:00] [ddtrace] [debug] - Flushing 0 traces, 1 spans
[2020-08-05T12:49:46+00:00] [ddtrace] [debug] - No finished traces to be sent to the agent
[2020-08-05T12:49:46+00:00] [ddtrace] [debug] - About to send trace(s) to the agent
[2020-08-05T12:49:46+00:00] [ddtrace] [debug] - circuit breaker status: closed => 1, total_failures => 0,consecutive_failures => 0, opened_timestamp => 0, last_failure_timestamp=> 0
[2020-08-05T12:49:46+00:00] [ddtrace] [debug] - Traces successfully sent to the agent
```

**PHP Info**

Output of `php -v`.

```
/tmp # php -v
PHP 7.4.8 (cli) (built: Jul  9 2020 23:47:36) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.8, Copyright (c), by Zend Technologies
    with blackfire v1.34.3~linux-musl-x64-non_zts74, https://blackfire.io, by Blackfire
    with ddtrace v0.47.1, Copyright Datadog, by Datadog
```

Output of `php -m`.

```
/tmp # php -m
[PHP Modules]
apcu
bcmath
blackfire
bz2
calendar
Core
ctype
curl
date
ddtrace
dom
fileinfo
filter
ftp
hash
iconv
igbinary
intl
json
libxml
maxminddb
mbstring
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
readline
redis
Reflection
session
SimpleXML
sodium
SPL
sqlite3
standard
tokenizer
uuid
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache
blackfire
ddtrace
```

**OS Info**

Output of `cat /etc/os-release | grep -E "(NAME)|(VERSION)"`.

```
/tmp # cat /etc/os-release | grep -E "(NAME)|(VERSION)"
NAME="Alpine Linux"
VERSION_ID=3.12.0
PRETTY_NAME="Alpine Linux v3.12"
```

**Output of dd-doctor**
- Deploy `dd-doctor` to your root folder `curl https://raw.githubusercontent.com/DataDog/dd-trace-php/master/src/dd-doctor.php -o <path-to-webroot>/<some-random-name>.php`
- Access it at `http://your-host/<some-random-name>.php`
- Paste here the output
- Remember to remove the file `<path-to-webroot>/<some-random-name>.php` when you are done.

In web mode.

```
DataDog trace extension verification

- PHP version and SAPI                               [7.4.8 - fpm-fcgi]
- ddtrace extension installed                        [OK]
- ddtrace version (installed)                        [0.47.1]
- ddtrace version (const)                            [0.47.1]
- ddtrace version (userland)                         [0.47.1]
- ddtrace versions in sync                           [OK]
- dd_trace() function available                      [OK]
- dd_trace_env_config() function available           [OK]
- ddtrace.request_init_hook set                      [OK]
- ddtrace.request_init_hook reachable                [OK]
- ddtrace.request_init_hook has run                  [FAIL]
- 'open_basedir' allows access to '/proc/self/'      [OK]
- IntegrationsLoader exists                          [FAIL]
- DDTrace\Tracer class exists                        [FAIL]
- Background sender is enabled?                      [NO]
  > You can enable the background sender via
    DD_TRACE_BETA_SEND_TRACES_VIA_THREAD=true
- Opcache installed:                                 [YES]
  > opcache.enable = 1
  > opcache.enable_cli = 1
  > opcache.use_cwd = 1
  > opcache.validate_timestamps = 
  > opcache.validate_permission = 
  > opcache.validate_root = 
  > opcache.dups_fix = 
  > opcache.revalidate_path = 
  > opcache.log_verbosity_level = 1
  > opcache.memory_consumption = 268435456
  > opcache.interned_strings_buffer = 32
  > opcache.max_accelerated_files = 50000
  > opcache.max_wasted_percentage = 0.1
  > opcache.consistency_checks = 0
  > opcache.force_restart_timeout = 180
  > opcache.revalidate_freq = 2
  > opcache.preferred_memory_model = 
  > opcache.blacklist_filename = 
  > opcache.max_file_size = 0
  > opcache.error_log = 
  > opcache.protect_memory = 
  > opcache.save_comments = 1
  > opcache.enable_file_override = 
  > opcache.optimization_level = 2147401727
  > opcache.lockfile_path = /tmp
  > opcache.file_cache = /opcache
  > opcache.file_cache_only = 
  > opcache.file_cache_consistency_checks = 1
  > opcache.file_update_protection = 0
  > opcache.opt_debug_level = 0
  > opcache.restrict_api = 
  > opcache.huge_code_pages = 
  > opcache.preload = 
  > opcache.preload_user = 
- Configured Agent host                              [datadog-agent]
- Configured Agent port                              [8126]
- Agent can receive traces                           [OK]
```

In CLI mode, I have several `[FAIL]` status, do you know why ?

```
/tmp # DD_TRACE_CLI_ENABLED=true php dd-doctor.php
WARNING: Script is running from the CLI SAPI.
         Please run this script from your web browser.

DataDog trace extension verification

- PHP version and SAPI                               [7.4.8 - cli]
- ddtrace extension installed                        [OK]
- ddtrace version (installed)                        [0.47.1]
- ddtrace version (const)                            [0.47.1]
- ddtrace version (userland)                         [0.47.1]
- ddtrace versions in sync                           [OK]
- dd_trace() function available                      [OK]
- dd_trace_env_config() function available           [OK]
- ddtrace.request_init_hook set                      [OK]
- ddtrace.request_init_hook reachable                [OK]
- ddtrace.request_init_hook has run                  [FAIL]
- 'open_basedir' allows access to '/proc/self/'      [OK]
- IntegrationsLoader exists                          [FAIL]
- DDTrace\Tracer class exists                        [FAIL]
- Background sender is enabled?                      [NO]
  > You can enable the background sender via
    DD_TRACE_BETA_SEND_TRACES_VIA_THREAD=true
- Opcache installed:                                 [YES]
  > opcache.enable = 1
  > opcache.enable_cli = 1
  > opcache.use_cwd = 1
  > opcache.validate_timestamps =
  > opcache.validate_permission =
  > opcache.validate_root =
  > opcache.dups_fix =
  > opcache.revalidate_path =
  > opcache.log_verbosity_level = 1
  > opcache.memory_consumption = 268435456
  > opcache.interned_strings_buffer = 32
  > opcache.max_accelerated_files = 50000
  > opcache.max_wasted_percentage = 0.1
  > opcache.consistency_checks = 0
  > opcache.force_restart_timeout = 180
  > opcache.revalidate_freq = 2
  > opcache.preferred_memory_model =
  > opcache.blacklist_filename =
  > opcache.max_file_size = 0
  > opcache.error_log =
  > opcache.protect_memory =
  > opcache.save_comments = 1
  > opcache.enable_file_override =
  > opcache.optimization_level = 2147401727
  > opcache.lockfile_path = /tmp
  > opcache.file_cache = /opcache
  > opcache.file_cache_only =
  > opcache.file_cache_consistency_checks = 1
  > opcache.file_update_protection = 0
  > opcache.opt_debug_level = 0
  > opcache.restrict_api =
  > opcache.huge_code_pages =
  > opcache.preload =
  > opcache.preload_user =
- Configured Agent host                              [datadog-agent]
- Configured Agent port                              [8126]
- Agent can receive traces                           [OK]
```

**Comment by SammyK** [comment]
Hello @bgaillard. Thank you for such a detailed report.

There is [a bug in OPcache](https://bugs.php.net/bug.php?id=79825) that interferes with the tracer when `opcache.file_cache` is set. We've already submitted upstream bug patches (php/php-src#5836 & php/php-src#5871), but they have not been merged yet. If you disable the `opcache.file_cache` setting, does that solve the issue?

If that doesn't solve the issue:
1. Did you upgrade ddtrace from an existing install or did you install fresh?
2. With the `0.47.1` install, could you run `php --ri=ddtrace` and provide the output?

**Comment by bgaillard** [comment]
Hello @SammyK. Thanks for you reactivity on this issue.

> There is a bug in OPcache that interferes with the tracer when opcache.file_cache is set. We've already submitted upstream bug patches (php/php-src#5836 & php/php-src#5871), but they have not been merged yet. If you disable the opcache.file_cache setting, does that solve the issue?

Yes, I just checked it with the following command and now it worked. Thanks !

```
/tmp # DD_AGENT_HOST=datadog-agent DD_AGENT_PORT=8126 DD_TRACE_DEBUG=true DD_TRACE_ENABLED=true DD_TRACE_NO_AUTOLOADER=true DD_TRACE_CLI_ENABLED=true DD_TRACE_GENERATE_ROOT_SPAN=true DD_TRACE_AUTO_FLUSH_ENABLED=true DD_SERVICE=hello-world php -d opcache.file_cache=NULL test.php
[2020-08-05T15:02:38+00:00] [ddtrace] [debug] - New span test.php  recorded.
...
Hello world !
[2020-08-05T15:02:38+00:00] [ddtrace] [debug] - Flushing 1 traces, 2 spans
[2020-08-05T15:02:38+00:00] [ddtrace] [debug] - Encoding span '7229623110529105847' op: 'test.php' serv: 'hello-world' res: 'test.php' type 'cli'
[2020-08-05T15:02:38+00:00] [ddtrace] [debug] - Tags for span 7229623110529105847 'tag:chars_count' are: version:40,env:16,system.pid:4
[2020-08-05T15:02:38+00:00] [ddtrace] [debug] - About to send trace(s) to the agent
Successfully memoized Agent HTTP headers
```

I'll check with my colleagues if disabling this option is problematic for the CLI scripts we need to instrument.

**Comment by SammyK** [comment]
I'm glad to hear that solved the issue for you. :) We'll keep an eye out for the upstream patches to be merged and update this thread when they are rolled into a patch release of PHP.

**Comment by bgaillard** [comment]
Hi @SammyK, during our conversation we missed an other element. 

When I execute `DD_TRACE_CLI_ENABLED=true php dd-doctor.php` with version `0.47.1` of `ddtrace` I get the following `[FAIL]` states.

```
- ddtrace.request_init_hook has run                  [FAIL]
- IntegrationsLoader exists                          [FAIL]
- DDTrace\Tracer class exists                        [FAIL]
```

Do you know what's causing this ? If it's an other bug do you wish me to create a new issue ?

Thanks

**Comment by SammyK** [comment]
@bgaillard It looks like that might be an older version of `dd-doctor.php` giving false-negatives. If you use [the `dd-doctor.php` script from `0.47.1`](https://raw.githubusercontent.com/DataDog/dd-trace-php/0.47.1/src/dd-doctor.php), do you still see the failures?

**Comment by bgaillard** [comment]
@SammyK Yes, I already downloaded the last version.

I just tried again with versions `0.41.1`, `0.44.1` and `0.47.1` and found the error I have.

It appears that when the environment variable `DD_TRACE_ENABLED=false` then I get the following errors.

```
- ddtrace.request_init_hook has run                  [FAIL]
- IntegrationsLoader exists                          [FAIL]
- DDTrace\Tracer class exists                        [FAIL]
```

When the variable is `true` everything's OK.

Don't know if it's normal or not but the error about a missing class `DDTrace\Tracer` was very confusing for me and I believed I had a very critical problem.

Perhaps `dd-doctor.php` could be improved a little to display most important `DD_*` environment variables and add much more details about errors ?

**Comment by SammyK** [comment]
Thanks for the feedback @bgaillard. I agree that the diagnostics from `dd-doctor.php` can show confusing results with certain configurations. Since version `0.47` of ddtrace we started phasing out `dd-doctor.php` in favor of a richer `phpinfo()` section with more comprehensive and less noisy diagnostics. It can also be accessed from `php --ri=ddtrace`. If the `Diagnostics` section passes, everything should be configured properly.

**Comment by SammyK** [comment]
Hey @bgaillard! The upstream patch for the `opcache.file_cache` issue (php/php-src#5836) was merged into the PHP 7.3 and PHP 7.4 branches and should be rolled into the next patch releases (7.3.23 & 7.4.11).

**Comment by bwoebi** [comment]
Given this seems to have been resolved and released for a while, I'm closing the issue now.
